docs/proposals/20210222-machineadm.md

```diff
@@ -0,0 +1,835 @@
+	MachineBootstrapConfigSpec `json:",inline"`
+	// PluginTemplateRefs is a list of machine bootstrap plugin template refs
+	// which will be instantiated for each machine.
+	PluginTemplateRefs []corev1.ObjectReference `json:"pluginTemplateRefs,omitempty"`
+}
+
+// MachineBootstrapConfigStatus defines the observed state of KubeadmConfig
+type MachineBootstrapConfigStatus struct {
+	// Ready indicates the BootstrapData field is ready to be consumed
+	Ready bool `json:"ready,omitempty"`
+
+	// DataSecretName is the name of the secret that stores the bootstrap data script.
+	// +optional
+	DataSecretName *string `json:"dataSecretName,omitempty"`
+
+	// FailureReason will be set on non-retryable errors
+	// +optional
+	FailureReason string `json:"failureReason,omitempty"`
+
+	// FailureMessage will be set on non-retryable errors
+	// +optional
+	FailureMessage string `json:"failureMessage,omitempty"`
+
+	// ObservedGeneration is the latest generation observed by the controller.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// Conditions defines current service state of the KubeadmConfig.
+	// +optional
+	Conditions clusterv1.Conditions `json:"conditions,omitempty"`
+}
+
+// PluginStatus defines the status of a machineadm plugin
+type PluginStatus struct {
+	// FailureReason will be set on non-retryable errors
+	// +optional
+	FailureReason string `json:"failureReason,omitempty"`
+	// FailureMessage will be set on non-retryable errors
+	// +optional
+	FailureMessage string `json:"failureMessage,omitempty"`
+	// Conditions defines current service state of the cluster.
+	// +optional
+	Conditions Conditions `json:"conditions,omitempty"`
+	// Ready defines if the control plane is ready.
+	// +optional
+	Ready         bool     `json:"ready,omitempty"`
+	SecretName    string   `json:"secretName,omitempty"`
+	BeforePlugins []string `json:"beforePlugins,omitempty"`
+	AfterPlugins  []string `json:"afterPlugins,omitempty"`
+}
+```
+
+#### Serialized data format
+
+`machineadm_config.yaml` will be a multi-part YAML docment supporting multiple
+data types, read and processed in order.
+
+##### Remote configuration
+
+```go
+// RemoteConfig is a resource for fetching a remote machine configuration
+type RemoteConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec RemoteConfigSpec `json:"spec,omitempty"`
+}
+
+type RemoteConfigSpec struct {
+	// Provider is the infrastructure provider plugin that fetches the remote config.
+	// This must be present on the host as
+	// machineadm-remote-configuration-provider-x
+	// +kubebuilder:validation:Required
+	Provider string `json:"provider"`
+	// URI is the location of the provider-specific remote config for retrieval
+	// s3://us-west-2.cluster-api.contoso.com/bootstrap/test-cluster-machine-ufjdhad.yaml
+	// +kubebuilder:validation:Required
+	URI string `json:"uri"`
+}
+```
+
+##### Encrypted configuration
```

Comment:
See above, the `machineadm_config.yaml` can be encrypted using sops, and then decrypted inline by `machineadm` there should not be any need to configure decryption parameters and can be inferred from the sops envelope

Reply:
I +1 here. I would like to keep encryption configuration to a minimum and sops _seems_ like a good fit to me.

(Disclaimer: not an expert on this)